Schema Manager retry for Validating Version Compatibility

abdullah248 opened the pull request:
## Description
Add retry when validating version compatibility as there are cases where a retry would succeed due to ttl in instance schema db.


## [Semver Change](https://github.com/microsoft/healthcare-shared-components/blob/main/docs/Versioning.md)
Patch

wsugarman commented on src/Microsoft.Health.SqlServer/Features/Schema/Manager/SqlSchemaManager.cs:
```diff
@@ -119,7 +119,18 @@ public virtual async Task ApplySchema(MutuallyExclusiveType type, bool force = f
+            // TTL in instance schema table can cause this call to fail on first attempt so allow for retries
+            attemptCount = 1;
+            await Policy.Handle<SchemaManagerException>()
+            .WaitAndRetryAsync(
```
nit: indent the methods by 1

wsugarman commented on src/Microsoft.Health.SqlServer/Features/Schema/Manager/SqlSchemaManager.cs:
```diff
@@ -119,7 +119,18 @@ public virtual async Task ApplySchema(MutuallyExclusiveType type, bool force = f
 
-            await ValidateVersionCompatibility(availableVersions[^1].Id, token).ConfigureAwait(false);
+            // TTL in instance schema table can cause this call to fail on first attempt so allow for retries
+            attemptCount = 1;
```
The `onRetry` delegate is providing this information; you don't need to track it yourself

wsugarman commented on src/Microsoft.Health.SqlServer/Features/Schema/Manager/SqlSchemaManager.cs:
```diff
@@ -119,7 +119,18 @@ public virtual async Task ApplySchema(MutuallyExclusiveType type, bool force = f
+            await Policy.Handle<SchemaManagerException>()
+            .WaitAndRetryAsync(
+                retryCount: RetryAttempts,
+                sleepDurationProvider: (retryCount) => TimeSpan.FromSeconds(60),
```
nit you don't need `()` for single parameter delegates

wsugarman commented on src/Microsoft.Health.SqlServer/Features/Schema/Manager/SqlSchemaManager.cs:
```diff
@@ -119,7 +119,18 @@ public virtual async Task ApplySchema(MutuallyExclusiveType type, bool force = f
+                sleepDurationProvider: (retryCount) => TimeSpan.FromSeconds(60),
+                onRetry: (exception, retryCount) =>
+                {
+                    _logger.LogError(exception, "Attempt {Attempt} of {MaxAttempts} of validating version compatiblity", attemptCount++, RetryAttempts);
```
Should this be `Warning` instead?

wsugarman commented on src/Microsoft.Health.SqlServer/Features/Schema/Manager/SqlSchemaManager.cs:
```diff
@@ -119,7 +119,18 @@ public virtual async Task ApplySchema(MutuallyExclusiveType type, bool force = f
+                retryCount: RetryAttempts,
+                sleepDurationProvider: (retryCount) => TimeSpan.FromSeconds(60),
+                onRetry: (exception, retryCount) =>
+                {
```
nit: use expression over block (which I guess is a type of expression) for single line delegates

wsugarman commented on src/Microsoft.Health.SqlServer/Features/Schema/Manager/SqlSchemaManager.cs:
```diff
@@ -119,7 +119,18 @@ public virtual async Task ApplySchema(MutuallyExclusiveType type, bool force = f
+                {
+                    _logger.LogError(exception, "Attempt {Attempt} of {MaxAttempts} of validating version compatiblity", attemptCount++, RetryAttempts);
+                })
+            .ExecuteAsync(token => ValidateVersionCompatibility(availableVersions[^1].Id, token), token)
```
nit: rename the token to `t` in `ExcecuteAsync` to avoid the confusion

wsugarman reviewed (commented):
I see a number of my comments also pertain to the existing "retryable" code too

abdullah248 commented:
Yeah I had reused the exisiting code. I updated all the existing ones and my changes as well. Also reason for previously they were using `attempt` to track number of retries is because they were using the wrong delegate which was actually providing the sleep duration, not the number of retries. I updated to use the right delegate which includes the retryCount. 

`WaitAndRetryAsync(int retryCount, Func<int, TimeSpan> sleepDurationProvider, Action<Exception, TimeSpan, int, Context> onRetry)`
